examples/src/test/java/jdocs/kafka/KafkaDocExample.java

```diff
@@ -4,9 +4,213 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+//#imports
+import akka.kafka.ConsumerSettings;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+//#imports
+
+import static jdocs.jpa.HibernateSessionProvider.HibernateJdbcSession;
+
 public interface KafkaDocExample {
 
   //#todo
   // TODO
   //#todo
+
+  //#handler
+  public class WordCountHandler extends JdbcHandler<ConsumerRecord<String, String>, HibernateJdbcSession> {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final ProjectionId projectionId;
+
+    private final Map<String, Integer> state = new HashMap<>();
+
+    public WordCountHandler(ProjectionId projectionId) {
+      this.projectionId = projectionId;
+    }
+
+    @Override
+    public void process(HibernateJdbcSession session, ConsumerRecord<String, String> envelope) {
+      String word = envelope.value();
+      int newCount = state.getOrDefault(word, 0) + 1;
+      logger.info(
+          "{} consumed from topic/partition {}/{}. Word count for [{}] is {}",
+          projectionId,
+          envelope.topic(),
+          envelope.partition(),
+          word,
+          newCount);
+
+      state.put(word, newCount);
+    }
+  }
+  //#handler
+
+  //#wordSource
+  public class WordEnvelope {
+    public final Long offset;
+    public final String word;
+
+    public WordEnvelope(Long offset, String word) {
+      this.offset = offset;
+      this.word = word;
+    }
+  }
+
+  class WordSource extends SourceProvider<Long, WordEnvelope> {
+
+    private final Source<WordEnvelope, NotUsed> src = Source.from(
+        Arrays.asList(new WordEnvelope(1L, "abc"), new WordEnvelope(2L, "def"), new WordEnvelope(3L, "ghi"), new WordEnvelope(4L, "abc")));
+
+    @Override
+    public CompletionStage<Source<WordEnvelope, ?>> source(Supplier<CompletionStage<Optional<Long>>> offset) {
+      return offset.get().thenApply(o -> {
+        if (o.isPresent())
+          return src.dropWhile(envelope -> envelope.offset <= o.get()).throttle(1, Duration.ofSeconds(1));
+        else
+          return src.throttle(1, Duration.ofSeconds(1));
+      });
+    }
+
+    @Override
+    public Long extractOffset(WordEnvelope envelope) {
+      return envelope.offset;
+    }
+
+  }
+  //#wordSource
+
+  //#wordPublisher
+  class WordPublisher extends JdbcHandler<WordEnvelope, HibernateJdbcSession> {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final String topic;
+    private final SendProducer<String, String> sendProducer;
+
+    public WordPublisher(String topic, SendProducer<String, String> sendProducer) {
+      this.topic = topic;
+      this.sendProducer = sendProducer;
+    }
+
+    @Override
+    public void process(HibernateJdbcSession session, WordEnvelope envelope) throws Exception {
```

Comment:
same here, at-least-once with async handler. No?

Reply:
yes, publishing with this handler should at-least-once, and there is a FIXME for that below